The Dungeon Master

EddieJ03 opened the pull request:
- [x] Give the Dungon Master Top Down View of Maze
    - [x] Create DungeonMaster class that extends Creature
    - [x] Remove ceiling & turn off gravity AND jump in servergamestate.cpp if DungeonMaster
    - [x] Set `player->physics.shared.corner` in server.cpp if DungeonMaster to have a higher y-position
    - [x] Extend and override update method in Camera to lock DungeonMaster (change pitch)
- [x] Bound Dungeon Master WASD Movement
- [x] Be able to zoom in/out (currently using keys I for in and O for out)
- [x] What if Dungeon Master drops and tries to reconnect? - updated SessionEntry to have a is_dm boolean
- [x] Add trap placement
    - [x] Finish trap placement for ALL traps
- [x] Add trap selection - use hotbar like for Player items
    - [x] Finish hotbar UI for ALL traps
- [x] Add max trap, trap placement cooldown, and trap expiration timer
- [x] Hide cursor, bring in crosshair for trap placement
- [ ] mouse left-click to bring up crosshair, mouse right-click to hide crosshair
- [x] Shift to zoom in, space to zoom out 
- [x] Send DM WASD movements to server to recalculate world coordinate 

BUGS TO FIX:
- [x] Fix trap placement bug
- [x] Fix wall highlight bug


Miscellaneous Optimizations:
- [x] Optimize sending mouse xpos and ypos
- [x] Ignore unmovable objects when sending
- [x] Ignore spaces as empty floor creation

EddieJ03 committed:
some initial DM code

EddieJ03 committed:
Merge branch 'dev' of https://github.com/ucsd-cse125-sp24/group3 into 88-the-dungeon-master

EddieJ03 committed:
finally have a dunegon master view and can move WASD (but no jump)

EddieJ03 committed:
Dungeon Master View (First Player To Enter is DM)

EddieJ03 committed:
SO MANY DEV CHANGES YAAAAAAA

EddieJ03 committed:
bug fixes & updates

EddieJ03 committed:
dungeon master is bounded

EddieJ03 committed:
satisfy linter

EddieJ03 committed:
satisfy linter

EddieJ03 committed:
if DM dropped, reconnect as DM

EddieJ03 committed:
pulled in dev changes

EddieJ03 committed:
do not render inventory GUI stuff if DM

EddieJ03 committed:
Turn off 1,2,3,4,q,e if DM

EddieJ03 committed:
DM zoom in and out done

EddieJ03 committed:
restrict Y zoom in/out range and have field in config to toggle btwn DM mode and player mode

EddieJ03 committed:
Add in config.json

EddieJ03 committed:
merged in dev changes

EddieJ03 committed:
remove duplicate SpikeTrap

EddieJ03 committed:
Merge branch 'dev' of https://github.com/ucsd-cse125-sp24/group3 into 88-the-dungeon-master

EddieJ03 committed:
new dev changes

EddieJ03 committed:
Delete y-pos print for DM and add out/ to .gitignore

EddieJ03 committed:
DM world position click

EddieJ03 committed:
Merge branch 'dev' of https://github.com/ucsd-cse125-sp24/group3 into 88-the-dungeon-master

EddieJ03 committed:
finally got the right grid cell on click!

EddieJ03 committed:
get in some initial highlight backbone code

EddieJ03 committed:
can now click floor and highlight red (not walls though :()

EddieJ03 committed:
DM can spam floorspike traps now

EddieJ03 committed:
pulled in recent dev changes

EddieJ03 committed:
we now have HOHPOC, or more commonly referred to as: highlight on hover, place on click

EddieJ03 committed:
more refined highlighting & trap placement

EddieJ03 committed:
Cap number of trap placements by DM

EddieJ03 committed:
DM Trap Inventory Hotbar In Place!

EddieJ03 committed:
Got walls to be highlighted (some inconsistencies though, need debugging)

EddieJ03 committed:
disappearing traps are in!

EddieJ03 committed:
Improved DM Hotbar

EddieJ03 committed:
cooldown + a weird bug

Tyler-Lentz committed:
fix troll seg fault :pray:

EddieJ03 committed:
Skip floor creation if CellType is OutsideTheMaze

EddieJ03 committed:
only update movable objects

EddieJ03 committed:
some bug fixes

EddieJ03 committed:
merged in dev

EddieJ03 committed:
compilation fixes

EddieJ03 committed:
fix player ceiling render

EddieJ03 committed:
DM movement speedup

EddieJ03 committed:
DM selection bug

EddieJ03 committed:
crosshair DM

EddieJ03 committed:
DM has torchlights

EddieJ03 committed:
turn off torchlights for DM

EddieJ03 committed:
have highlight moving when WASD, but need to fix the highlight showing

gilkeidar committed:
fix incorrect placement because of second setWorldPos call & fixed NAN bug in set world pos

EddieJ03 committed:
have highlight (mostly) keepup when WASD is held

EddieJ03 committed:
better DM highlight on WASD movement

EddieJ03 committed:
some more movement highlight fixes

EddieJ03 committed:
only send one TrapPlacement event if multiple WASD pressed and/or zoom in/out

EddieJ03 committed:
satisfy linter

EddieJ03 committed:
dang it I missed one. Also i forgot to only send WASD movement if p is also pressed.

Tyler-Lentz committed:
in progress fucked up merge

Tyler-Lentz committed:
fix crash on place trap for DM

Tyler-Lentz committed:
fix ceiling not rendering for player

Tyler-Lentz committed:
YIPEE!

Tyler-Lentz committed:
fix lint

Tyler-Lentz committed:
completely necessary merge :smile:

Tyler-Lentz reviewed (approved):
Compiled and Ran on my Nintendo Entertainment System. Nice PR!

Tyler-Lentz committed:
fix wall and werror breaking windows

Tyler-Lentz committed:
partial performance fixes: misc tiny optimizations & no longer place floor under walls

Tyler-Lentz committed:
reenable first player dm

Tyler-Lentz committed:
Various fixes / optimizations

1. Made client not render internal walls
2. Adjusted networking code to not allocate heap memory every received packet and instead use static buffer
3. Fix item inventory not rendering correctly for non-DM players
4. Cleaned up classification of "walllike" cells for deciding whether or not to create floors

Tyler-Lentz committed:
general cleanup & remove extraneous comments

Tyler-Lentz committed:
more misc fixes and touchups

1. fix accidental regression on placing floors outside of the maze
2. slightly optimize render loop

EddieJ03 committed:
fix wall color for DM, revert to one big big floor to improve performance. RIP many floors

Tyler-Lentz committed:
Fix stupid ass networking "optimization"

Tyler-Lentz committed:
fix gui & add orb position display

gilkeidar committed:
dm wall textures

Tyler-Lentz committed:
fix item drop in walls
